src/keyboardNavigation.ts

```diff
@@ -0,0 +1,82 @@
+import { isDeepestVisible, isAncestor, isEmptyObject } from './utils';
+
+function findDeepestVisibleChild(node: HTMLElement): HTMLElement {
+    if (isDeepestVisible(node)) return node;
+    const deeperObject = <HTMLElement>node.querySelector('[part=object]');
+    const nextDeepest = getLastChild(deeperObject);
+    if (!nextDeepest) return node;
+    return findDeepestVisibleChild(nextDeepest);
+}
+
+function findNearestParentWithAdditionalLeaves(node: HTMLElement, initial = node): HTMLElement {
+    const ul = node.parentElement;
+    if (!ul) return node;
+    const parent = ul.parentElement;
+    if (!parent || !parent.parentElement) return initial;
+    const parentAndSiblings = [...parent.parentElement.children];
+    if (isLastChild(parent, parentAndSiblings)) {
+        return findNearestParentWithAdditionalLeaves(parent, initial);
+    } else {
+        return getNextSibling(parent, parentAndSiblings);
+    }
+}
+
+export const navigateUp = (path: string, shadowRoot: ShadowRoot): void => {
+    const currentFocus = <HTMLElement>shadowRoot.querySelector(`[data-path="${path}"]`);
+    if (!currentFocus || !currentFocus.parentNode) return;
+    const currentFocusAndSiblings = [...currentFocus.parentNode.children];
+    const previousSibling = getPreviousSibling(currentFocus, currentFocusAndSiblings);
```

Comment:
related to my last comment about moving out the grandparent selection from the sibling selection: i'm struck with the feeling that lines 28-34 here would make a lot more sense if the "sibling" case and the "grandparent" case were handled completely separately. as-is it's kinda tough to think through and give feedback on

Reply:
Mhh I'm not sure I fully understand what you mean here. My thinking is that we should be applying the same logic to see if its a valid focus element or not, no matter if the element is a grandparent or sibling.

Reply:
i guess what i mean is, i'm looking at the subsequent logic and i'm not sure all of it applies. like if `getPreviousSibling` grabs a sibling, it's definitely not a shadow root, nor an ancestor of the old path (unless there's an unlucky string match, i think). but staring at it harder, i think i'm convinced.

now that i've stared sufficiently, i have different feedback lol: `!x && !y ? a : b` is equivalent to `!(x || y) ? a : b`, and `x || y ? b : a`, the last of which i think actually helps readability: "is this an ancestor of the current path, or a node with no visible descendants? select it! otherwise, find its deepest descendant"